Try: Remove/refactor complex left/right block margins

jasmussen opened the pull request:
Every block, specifically `.block-editor-block-list__block`, has had paddings left and right to increase the hoverable area to invoke the up/down mover. This padding is then compensated for by a deeper container with negative left and right margins. 

Since the mover control is now only visible on-select, those negative margins are no longer necessary.

This PR refactors/removes those negative margins. This drastically simplifies the CSS of the block list component, and should make it easier for themes to get good results when they style the editor.

This is a work in progress. When done it needs to not land in 5.3, and it will need a dev-note when ready because unfortunately it is likely to break a number of existing editor styles. For example it breaks the off-set column and wide/full-wide styles in TwentyNineteen. The good news is that it should be easy to fix by removing a ton of hacky TwentyNineteen editor CSS.

Things to test when the PR becomes ready for review:

- Test primarily and first on a theme that does not style the editor.
- Test master and this branch and both should look identical.
- Test wide and fullwide and look for no horizontal scrollbars.
- Test every block that features nesting, notably Cover, Columns, Group.

While this PR is ready for review, it needs a lot of testing, and given its post-5.3 destination, I wouldn't call it urgent.

However the code quality is drastically improved, and there's lots of _red_ to be found, so it's definitely worth it.

jasmussen committed:
Try: Remove complex left/right block margins.

Those margins were created to accommodate a mover control that was available on hover. Given the mover control is now on-select only, this margin is no longer necessary.

This is a work in progress. When done it needs to not land in 5.3, and it will need a dev-note when ready.

youknowriad reviewed (commented):
Nice PR, seems to work well but when I tried the widgets screen, it was empty. I wonder if a rebase should fix it or maybe there's something else?

simison commented on assets/stylesheets/_variables.scss:
```diff
@@ -41,7 +41,7 @@ $shadow-modal: 0 3px 30px rgba($dark-gray-900, 0.2);
 // Editor Widths
 $sidebar-width: 280px;
-$content-width: 610px; // For the visual width, subtract 30px (2 * $block-padding + 2px borders). This comes to 580px, which is optimized for 70 characters.
+$content-width: 580px; // This is optimized for 70 characters.
```
Heads up that this needs rebasing after https://github.com/WordPress/gutenberg/pull/17883 :-)

youknowriad reviewed (approved):
LGTM

jasmussen commented:
Wow, that went easier than I'd expected :) 

Thank you for the review. I think I'll still want a few more eyes on it — it is unfortunately likely to require quite a few editor styles to need adjustments.

jasmussen commented:
Rebased. One of the reasons this needs more eyes is that it does affect editor styles. Here's TwentyNineteen:

![Screenshot 2019-10-29 at 14 43 36](https://user-images.githubusercontent.com/1204802/67772580-9b3d5780-fa5a-11e9-8979-2b0bbfd95542.png)

I'll try and prepare a patch for that, shouldn't be hard. But it's little things where the previous left and right margins were compensated for, in a way that's no longer necessary.

mapk reviewed (approved):
Looks good. I tested on FF with several blocks.

I used to see extra padding in the Inspector like this:

![Screen Shot 2019-10-30 at 8 24 10 AM](https://user-images.githubusercontent.com/617986/67872168-d1c4b200-faee-11e9-907f-847bbae7e8f0.png)

But now it all looks good! Thanks for simplifying this!

![padding](https://user-images.githubusercontent.com/617986/67872198-dc7f4700-faee-11e9-94d9-afa32fb43e52.gif)

jasmussen commented:
Thanks Mark for testing, and thanks everyone else for also testing. It's important to get many eyes on this, not just to test that it's good, but also to raise awareness of the change happening, so that if/when subsequent issues pop up, there's a chance you know why they're happening and where they originated :) 

Anyone know why the tests are failing? Given the depth of the changes it could definitely be legitimate failures, so would appreciate insights by anyone who has the time to look.

youknowriad commented:
I think the test failure is unrelated, I restarted the job to see.

jasmussen commented:
Rebased and squashed, just to keep it simpler to shepard along. 

I've also tested image floats, and they work well here.

jasmussen commented:
I restored a rule for the title block that I had deemed unnecessary. But resting it fixes the issue with TwentyNineteen, and therefore very probably a ton of other themes, so in doing so it makes this PR more resilient and less risky. 

I also included a todo in the CSS: https://github.com/WordPress/gutenberg/pull/17877/commits/c10412751233de429b1cdb876241e7baea6169d1 — we will revisit that area regardless if/when the Title becomes a block.

I did a bunch of testing in themes, and it's looking good:

Astra: 

![Screenshot 2019-11-01 at 10 15 02](https://user-images.githubusercontent.com/1204802/68014928-26079780-fc91-11e9-98ca-4d7edf76335d.png)

Vanilla:

![Screenshot 2019-11-01 at 10 15 19](https://user-images.githubusercontent.com/1204802/68014946-2f90ff80-fc91-11e9-8145-611744d42813.png)

A bunch of the Twenty themes:

![Screenshot 2019-11-01 at 10 15 30](https://user-images.githubusercontent.com/1204802/68014951-361f7700-fc91-11e9-8b5a-7a1de45b4c3c.png)

![Screenshot 2019-11-01 at 10 15 43](https://user-images.githubusercontent.com/1204802/68014953-3750a400-fc91-11e9-824c-ba83547aa48d.png)

![Screenshot 2019-11-01 at 10 15 54](https://user-images.githubusercontent.com/1204802/68014958-391a6780-fc91-11e9-9e42-999a79e7ec78.png)

![Screenshot 2019-11-01 at 10 16 32](https://user-images.githubusercontent.com/1204802/68014961-3a4b9480-fc91-11e9-8bb0-104be0f75ed3.png)

TwentyTwenty, notably, looks delightful:

![Screenshot 2019-11-01 at 10 17 32](https://user-images.githubusercontent.com/1204802/68014970-40417580-fc91-11e9-8fd6-466a9d5f0f21.png)

With all this, I'm feeling much more calm about merging this. It still needs a dev note, and it would still be good to go out in the next Gutenberg plugin so themers and block devs can be aware of this early.

jasmussen commented:
Rebased, squashed, and would appreciate insights on when's a good time to merge this.

For the record I'm seeing two issues this PR has caused for TwentyNineteen:

The line before the Title isn't flush:
<img width="366" alt="Screenshot 2019-11-04 at 08 48 08" src="https://user-images.githubusercontent.com/1204802/68105825-01edc580-fee0-11e9-84c6-19d0b055109c.png">

Fullwide is broken:
<img width="888" alt="Screenshot 2019-11-04 at 08 48 16" src="https://user-images.githubusercontent.com/1204802/68105827-02865c00-fee0-11e9-9fb1-e79a6d09c813.png">

This is only in TwentyNineteen, not in any of the other ones, and it's becasue that theme compensated for the margins this PR removes. I will push a PR to twentyninteen once this one is merged.

jasmussen commented:
Rebased this again, also to see if that makes the tests pass. If not, I could use some help to figure out if there's a legitimate test failure. @gziolo if you have time?

gziolo commented:
> Rebased this again, also to see if that makes the tests pass. If not, I could use some help to figure out if there's a legitimate test failure. @gziolo if you have time?

Travis seems to be happy about your work. It should be given that this PR adds only visual changes applied through CSS styles. I guess this is good to go based on previous approvals.

jasmussen commented:
Woop! I guess it's merge time. 

It will require a patch to TwentyNineteen, so I'll probably put this on my schedule for tomorrow, merge then patch twentynineteen. Thanks!

jasmussen commented:
I have created https://core.trac.wordpress.org/ticket/48526 to address the issues introduced in TwentyNineteen by this patch. 

The dev note for this patch is in https://github.com/WordPress/gutenberg/pull/17877#issuecomment-548043470.

I'm merging now and hopefully we can get the core patch in shortly after 5.3.

diegohaz commented on packages/block-editor/src/components/block-list/style.scss:
```diff
@@ -559,33 +557,16 @@
-				width: 100%;
-			}
+			margin-left: -$block-padding - $block-padding - $block-side-ui-width - $border-width - $border-width;
+			margin-right: -$block-padding - $block-padding - $block-side-ui-width - $border-width - $border-width;
```
I think there may be a styling issue introduced by these changes.

1. Create a Table block
2. Change its alignment to full width
3. Check mobile screen

![image](https://user-images.githubusercontent.com/3068563/68673677-97c6c780-0554-11ea-8a31-836c9d8c6e84.png)
